File: app/src/main/java/com/nwise/mvptemplate/ui/base/BasePresenter.java
package com.nwise.mvptemplate.ui.base;

import com.nwise.mvptemplate.di.components.ActivityComponent;
import io.reactivex.disposables.CompositeDisposable;

public abstract class BasePresenter<V> extends AbstractPresenter<V> {

    private boolean isFirstTime = true;
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();

    @Override
    public void onViewAttached(V view) {
        super.onViewAttached(view);
        if (isFirstTime) {
            isFirstTime = false;
            onViewAttachedForFirstTime(view);
        }
    }

    protected void onViewAttachedForFirstTime(V view) {
    }

    protected void addDisposable(CompositeDisposable disposable) {
        compositeDisposable.add(disposable);
    }

    @Override
    public void onViewDetached() {
        if (compositeDisposable != null && !compositeDisposable.isDisposed())
            compositeDisposable.dispose();
        super.onViewDetached();
    }
}

File: app/src/main/java/com/nwise/mvptemplate/ui/base/BasePresenter.kt
package com.nwise.mvptemplate.ui.base

import io.reactivex.disposables.CompositeDisposable

/**
 * Created by Sepideh Vatankhah on 01.08.2019.
 * sun.vatankhah@gmail.com
 * https://github.com/sepidevatankhah
 */
abstract class BasePresenter<V> : AbstractPresenter<V>() {

    private var isFirstTime = true
    private val compositeDisposable = CompositeDisposable()

    override fun onViewAttached(view: V) {
        super.onViewAttached(view)
        if (isFirstTime) {
            isFirstTime = false
            onViewAttachedForFirstTime(view)
        }
    }

    protected fun onViewAttachedForFirstTime(view: V) {}

    protected fun addDisposable(disposable: CompositeDisposable) {
        compositeDisposable.add(disposable)
    }

    override fun onViewDetached() {
        if (compositeDisposable != null && !compositeDisposable.isDisposed)
            compositeDisposable.dispose()
        super.onViewDetached()
    }
}
